### ISSUE TITLE
[RFC] Define Interfaces Using WIT

### ISSUE THREAD

**Issue by autodidaddict** [wasmCloud/wasmCloud#336]
## Summary
This request for comment aims to collect feedback and commentary on the idea to transition wasmCloud from using [Smithy](https://aws.amazon.com/blogs/developer/introducing-smithy-idl-2-0/) to define all interfaces to using [wit](https://github.com/WebAssembly/component-model/blob/main/design/mvp/WIT.md).

## Rationale
In the current wasmCloud ecosystem with the current state of tooling, creating actors and providers requires the use of **Smithy** to declare the interface used by wasm component actors and capability provider implementations. While Smithy may not be a closed system (it was created by Amazon), it is far from ubiquitous. Smithy doesn't enjoy the same kind of recognition and tooling ecosystem as [protocol buffers](https://github.com/protocolbuffers/protobuf) (aka **protobufs**), [Avro](https://avro.apache.org), or optimizations like [BSON](https://bsonspec.org) or [cap'n proto](https://capnproto.org). 

With the status quo, if I want to work with an existing interface like `wasmcloud:httpserver`, my actor code must use both guest and host generated code that comes from a Smithy definition. Our current code generation tools only support Rust and (for the most part) Go. It also requires that if we want to support a new target language for either side of the interface, we have to write our own code generation templates for that target language and wasmCloud developers must then use that generation. This kind of ecosystem management does not scale, nor is it remotely practical.

Another source of pain is that if I want to create a brand new interface, something like `custom:iot` for IoT devices, I have to write the interface in Smithy, and I am forced to implement my provider in one of the two languages supported by our Smithy code generation, and all actors using this interface must also be either Rust or Go. In every case, the code a developer creates must either rely directly on generated code during the build phase (e.g. `build.rs`), or it must rely on a library/crate that encapsulates the generated code (like [this](https://crates.io/crates/wasmcloud-interface-blobstore) or [this](https://crates.io/crates/wasmcloud-interface-httpserver)).

Last but not least, Smithy and its related tooling have no support at all for WebAssembly components, which locks the current IDL and tools out of this roadmap.

Summed up, this situation includes the following warning signs:
* Vendor lock-in (possible-but far from practical-to escape this lock-in)
* Somewhat niche community around Smithy
* Limited language choice and slow or non-existent addition of new language support
  * Requires tremendous effort from the maintainers to add language support
* No support for WebAssembly components

## Implementation Plan
As stated in the initial summary, the plan is to transition to **wit** for all interface definitions. The trick, as usual, is in how we implement this. It will require a progressive adoption that will follow the adoption and stability curves for wit and wasm components. This plan also has impact on the host, capability providers, and guests (actor components).

### Guest Wit with Wasifills
As mentioned at the beginning of this RFC, wasifills are the key piece of the implementation that makes all the puzzle pieces fit together. Remember that with WebAssembly components, the component can be written in any component-supporting language. This means that we can write Rust code that translates between discrete interface functions like `keyvalue`'s `get` and opaque binaries that the host never inspects. This translation happens in our (wasmCloud) code in a portable `.wasm` file, allowing developers to write their own code and have **_no tightly coupled dependencies on any wasmCloud libraries_**.

In both of the phases below, we should be able to make it so that `wash build` selects the appropriate wasifill `.wasm` file and links it, producing a module that "just works" with wasmCloud.

In an ideal world, since we can control the serialization of payloads, we should be able to use automatic code generation to produce wasifills for any wit contract on demand. 🤞🏻

#### Phase 1
The first phase will be to define the existing wasmCloud interfaces in terms of **wit**. This will be a breaking change in terms of the names and data types of the functions being used, but the API surface area should be about the same. Every function supported on today's Smithy interfaces will be available in the phase 1 interfaces. 

Using these new interfaces would involve the use of [wit-bindgen](https://github.com/bytecodealliance/wit-bindgen), and so guest language support would automatically be the set of all languages supported by **wit-bindgen**. The reason for this phase is that the specifications for the various wasi interfaces like `wasi-messaging` and `wasi-keyvalue` are still highly volatile.

```rust
wit_bindgen::generate!({
    world: "key-value",
    path: "../wit",
});

...

match keyvalue::get("something") {
 ...
}
```
This would reach out to the wasmCloud-specific **wit** and generate the appropriate code. There would no longer be any proprietary or vendor-locked code in any of the guest components. During this phase, the interfaces will appear similar but slightly different than the officially accepted wasi cloud interfaces. 

During this phase a wasifill would be linked to developer Wasm components to provide the right interface to wasmCloud hosts. Remember that the developer's code has no wasmCloud-specific code, the wasifill provides that translation layer.

#### Phase 2
During phase 2, we support the standard specifications for wasi cloud interfaces that are stable enough for use in real applications. Note that not all interfaces will be in the same phase at the same time: we can and will support both simultaneously. For example, during the `wash build` phase, it could select `keyvalue_wasifill_p1.wasm` or `keyvalue_wasifill_p2.wasm`. These wasifills could even be sitting in a registry or repository somewhere so that they could be pulled in by `wash` and not stored with the developer's project.

All it would take to change from phase 1 to phase 2 is changing the source of the wit world definition, and changing any code that needs to change as a result of the different interface. If we're lucky, most of the interfaces should migrate with minor impact. For example, the following phase 2 sample code might be almost indistinguishable from phase 1:

```rust
wit_bindgen::generate!({
    world: "wasi-keyvalue",
    path: "../wit",
});

...

match keyvalue::get("something") {
 ...
}
```

### Provider Implementations
Migrating capability providers to work with `wit` interfaces is actually a bit more difficult and requires more planning and forethought than with the guest modules. Today's capability providers use Smithy-generated serialization and de-serialization code (uses [messagepack](https://msgpack.org/index.html) right now). To create a provider for a given interface, the current process involves declaring a Rust dependency on the shared crate containing the various translation code.

Capability providers all handle a single wasmCloud function that accepts an invocation. Anything more strongly typed than that is generated on top of the lowest level primitive. To enable providers to continue to support bi-directional RPC with hosts, they will need to continue to support this mode of operation. The difference will be that instead of a provider relying on a shared library for its translation functionality, phase 1 and 2 providers will rely on wit-bindgen generated code and translation shims that extract payloads from invocations and invoke wit-style functions.

Instead of using a Smithy/weld shared crate, a wit-bindgen-based provider might look something like this:

```rust
wit_bindgen::generate!({
    path: "../wit",
    world: "pingpong",
});

struct Pingpong;
impl handler::Handler for Pingpong {
    fn on_receive(e: Event) -> Result<(), u32> {
     ...
    }
}
```
We should be able to generate the necessary "lifting" code to extract a payload from an invocation on the wire and invoke the appropriate strongly-typed function. It's unclear at this time, but there may be some things related to wasifill shims that we can do on the provider side, even though they aren't WASI-based (yet).

## Risks
This RFC makes some assumptions about the maturity and ease of use of certain tools. It assumes that we'll be able to make capability providers as loosely coupled from wasmCloud as guest components. In a worst case scenario, we might have to rely on shared code between the wasifill and the provider (which would still leave the guests totally uncoupled).

In other words, there's pretty low risk on the guest/actor side of the work in this RFC. There are some unknowns surrounding the provider side of this, but even that worst case scenario of manual translation layers isn't all that bad, and is still better than the status quo.

**Comment by autodidaddict** [comment]
I expect that @ricochet or @rvolosatovs might be able to provide some more detail around how the provider implementations might work, thus removing some of the unknowns from the risk section.

**Comment by aschrijver** [comment]
No technical feedback. First of all I'd like to applaud how you folks interact with the community :clap: 

I'm new to Wasm/WASI and doing a kind of technology survey, and thus represent the "average newcomer" to the field. There's a lot of moving parts in this ecosystem to wrap your head around, and - as [top HN comment](https://news.ycombinator.com/item?id=36127994) on WASIX announcement also states - on the WASI side documentation is.. discouraging an entry. After Wasm/WASI tech picture comes the question of "What vendors/projects are active?"

wasmCloud/Cosmonic are well-positioned in this regard. I like a lot the focus on minimal vendor lock-in and and standards-based techstack choices. Those are true USP's imho. But those building blocks add to the cognitive load of my tech survey. And I was like "Wait. Smithy?? How does that fit in? Where's WIT?" .. Yet more stuff to wrap your head around.

- The average newcomer / potential early adopter expects a workflow that matches Wasm/WASI vision.
- Smithy being temporary in the stack, an additional moving part to learn, is a barrier to entry on wasmCloud part.
- The longer Smithy is in the stack, the more people will need to transition away from it on the WIT-capable release.

My feedback would be to transition sooner rather than later, even if that means being more on the bleeding edge. Not doing so also constitutes something of a risk to the project by people moving elsewhere to a project perceived to be easier to familiarize with quickly.

**Comment by brooksmtownsend** [comment]
@autodidaddict I think this is above and beyond accepted and I propose we move this to an ADR to close this!

**Comment by connorsmith256** [comment]
Not stale, still just need to convert this to an ADR 🙂
